src/Toolkit/Toolkit/UI/Controls/Legend/LegendDataSource.cs

```diff
@@ -0,0 +1,541 @@
+                UpdateItemsSource();
+            }
+#endif
+            if (e.PropertyName == nameof(MapView.MapScale) && sender is MapView mv)
+            {
+                _currentScale = mv.MapScale;
+                MarkCollectionDirty();
+            }
+        }
+
+        private double _currentScale = double.NaN;
+
+        private List<object> BuildLegendList(IEnumerable<ILayerContent> layers)
+        {
+            if (layers == null)
+            {
+                return null;
+            }
+
+            List<object> data = new List<object>();
+            foreach (var layer in layers)
+            {
+                if (!layer.ShowInLegend || (!layer.IsVisible && _filterHiddenLayers))
+                {
+                    continue;
+                }
+
+                if (layer is Layer l)
+                {
+                    var state = _geoview.GetLayerViewState(l);
+                    if (state != null &&
+                        ((state.Status == LayerViewStatus.NotVisible && _filterHiddenLayers) ||
+                        (state.Status == LayerViewStatus.OutOfScale && _filterByVisibleScaleRange)))
+                    {
+                        continue;
+                    }
+                }
+                else if (layer is ILayerContent ilc)
+                {
+                    if (!ilc.IsVisible && _filterHiddenLayers)
+                    {
+                        continue;
+                    }
+                    else if (_filterByVisibleScaleRange && !double.IsNaN(_currentScale) && _currentScale > 0 && !ilc.IsVisibleAtScale(_currentScale))
+                    {
+                        continue;
+                    }
+                }
+
+                data.Add(layer);
+                if (!(layer is Layer) || ((Layer)layer).LoadStatus == LoadStatus.Loaded)
+                {
+                    // Generate the legend infos
+                    // For layers, we'll wait with entering here, until the GeoView decides to load them before generating the legend
+                    if (!_legendInfoTasks.ContainsKey(layer))
+                    {
+                        var task = LoadLegend(layer);
+                        _legendInfoTasks[layer] = task;
+                    }
+                    else
+                    {
+                        var task = _legendInfoTasks[layer];
+                        if (task.Status == TaskStatus.RanToCompletion)
+                        {
+                            var legends = _legendInfoTasks[layer].Result;
+                            data.AddRange(legends);
+                        }
+                    }
+                }
+
+                if (layer.SublayerContents != null)
+                {
+                    data.AddRange(BuildLegendList(_reverseLayerOrder ? layer.SublayerContents : layer.SublayerContents?.Reverse() )); // This might seem counter-intuitive, but sublayers are already top-to-bottom, as opposed to the layer collection
```

Comment:
Note: ArcGISMapImageLayer sublayers should not be reversed. It's different from other sublayers.